### ISSUE TITLE
Chroma Dependency Lock

### ISSUE THREAD

**Issue by MattGrohoNH** [langroid/langroid#396]
**Describe the bug**
The `chromadb` package is locked to version `0.3.21` and is beginning to become outdated. 

**To Reproduce**
Adding `langroid` as a dependency to any other Python package which depends on any other version of `chromadb`.

**Expected behavior**
Should be allowed to specify a particular version of Chroma `chromadb>=0.3.29`

**To Fix**
This line in `pyproject.toml`:
`chromadb = "0.3.21"`
should be replaced with:
`chromadb = "^0.3.21"`

**Comment by Mohannadcse** [comment]
@MattGrohoNH thanks for raising this issue. The current PyPi chroma version is `0.4.22`. I'm working on this, we need to make sure the new version doesn't cause any issue.

**Comment by pchalasani** [comment]
@Mohannadcse we can try to get  it to work with >= 0.3.29

**Comment by Mohannadcse** [comment]
tried the recent version `0.4.22` and doesn't work, 
the recent version that `poetry update` allows is `0.3.26`. But when I demanded this version using `pip install chromadb==0.3.29`, it showed me some conflict
```
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
langroid 0.1.153 requires chromadb==0.3.21, but you have chromadb 0.3.29 which is incompatible.
langroid 0.1.153 requires pydantic==1.10.11, but you have pydantic 1.10.13 which is incompatible.
chainlit 1.0.200 requires fastapi<0.109.0,>=0.100, but you have fastapi 0.85.1 which is incompatible.
Successfully installed chromadb-0.3.29 fastapi-0.85.1 starlette-0.20.4
```

**Comment by Mohannadcse** [comment]
> >= 0.3.29

so we definetly can't do this, because this will get version `0.4.22` installed.

**Comment by MattGrohoNH** [comment]
Hey @Mohannadcse thanks so much for your diligence with this issue!

I have since made a fork to test chromadb versioning and was able to get it working using: `chromadb = ">=0.3.21, <0.4.22"`

Let me know if this works as a temporary fix for `0.4.22`.

( We utilize the PersistentClient from chromadb which is missing in `0.3.21` )

**Comment by Mohannadcse** [comment]
Thanks @MattGrohoNH for the update and the details. 
Which Chromadb you are currently using after making this modification `chromadb = ">=0.3.21, <0.4.22"`?
We can update `pyproject.toml` to have this version range condition, while we work on the upgrade to version `0.4.22`.

**Comment by MattGrohoNH** [comment]
I'm currently using `0.4.21`, though and version greater than `0.4.14` should have the PersistentClient functionality.

**Comment by pchalasani** [comment]
@MattGrohoNH  we've updated to chroma 0.4.23, but note that it is now an "extra" dependency, which you have to install using 
```
pip install "langroid[chromadb]"
```
or equivalent.

**Comment by pchalasani** [comment]
Fixed in PR #415
